ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm

```diff
@@ -0,0 +1,778 @@
+#  );
+#}
+
+#use constant \%rule_constants;
+
+use base 'Exporter';
+
+my @rule_constants = ('LINE_FORMAT_PARAM',
+                   'LINE_FORMAT_ENVVAR',
+                   'LINE_FORMAT_XRDCFG',
+                   'LINE_FORMAT_XRDCFG_SETENV',
+                   'LINE_FORMAT_XRDCFG_SET',
+                   'LINE_VALUE_AS_IS',
+                   'LINE_VALUE_BOOLEAN',
+                   'LINE_VALUE_HOST_LIST',
+                   'LINE_VALUE_INSTANCE_PARAMS',
+                   'LINE_VALUE_ARRAY',
+                   'LINE_VALUE_HASH_KEYS',
+                   'LINE_VALUE_STRING_HASH',
+                   'LINE_VALUE_OPT_NONE',
+                   'LINE_VALUE_OPT_SINGLE',
+                   'LINE_FORMAT_DEFAULT',
+                   'LINE_QUATTOR_COMMENT',
+                   'LINE_OPT_DEF_REMOVE_IF_UNDEF',
+                   'LINE_OPT_DEF_ALWAYS_RULES_ONLY',
+                  );
+our @EXPORT_OK;
+our %EXPORT_TAGS;
+push @EXPORT_OK, @rule_constants;
+$EXPORT_TAGS{rule_constants} = \@rule_constants;
+
+
+# Backup file extension
+use constant BACKUP_FILE_EXT => ".old";
+
+
+
+# This function formats an attribute value based on the value format specified.
+#
+# Arguments :
+#        attr_value : attribue value
+#        line_fmt : line format (see LINE_FORMAT_xxx constants)
+#        value_fmt : value format (see LINE_VALUE_xxx constants)
+sub formatAttributeValue () {
+  my $function_name = "formatAttributeValue";
+  my $self = shift;
+
+  my $attr_value = shift;
+  unless ( defined($attr_value) ) {
+    $self->error("$function_name: 'attr_value' argument missing");
+    return 1;
+  }
+  my $line_fmt = shift;
+  unless ( defined($line_fmt) ) {
+    $self->error("$function_name: 'list_fmt' argument missing");
+    return 1;
+  }
+  my $value_fmt = shift;
+  unless ( defined($value_fmt) ) {
+    $self->error("$function_name: 'value_fmt' argument missing");
+    return 1;
+  }
+
+  $self->debug(2,"$function_name: formatting attribute value >>>$attr_value<<< (line fmt=$line_fmt, value fmt=$value_fmt)");
+
+  my $formatted_value;
+  if ( $value_fmt == LINE_VALUE_HOST_LIST ) {    
+    # Duplicates may exist as result of a join. Checkt it.
+    my @hosts = split /\s+/, $attr_value;
+    my %hosts;
+    for my $host (@hosts) {
+      unless ( exists($hosts{$host}) ) {
+        $hosts{$host} = "";
+      }
+    }  
+    my $formatted_value="";
+    for my $host (sort keys %hosts) {
+      $formatted_value .= "$host ";
+    }
+    # Some config files are sensitive to extra spaces : suppress trailing spaces
+    $formatted_value =~ s/\s+$//;
+    $self->debug(1,"Formatted hosts list : >>$formatted_value<<");
+
+  } elsif ( $value_fmt == LINE_VALUE_BOOLEAN ) {
+    if ( $attr_value ) {
+      $formatted_value = '"yes"';
+    } else {
+      $formatted_value = '"no"';
+    }
+
+  } elsif ( $value_fmt == LINE_VALUE_INSTANCE_PARAMS ) {
+    # Instance parameters are described in a nlist
+    if ( exists($attr_value->{logFile}) ) {
```

Comment:
there's no need for `exists`, is `$attr_value->{logFile}` is undef or empty string, you don't this anyway. so simply
```perl
$formatted_value .= " -l $attr_value->{logFile}" if ($attr_value->{logFile});
```